_posts/2019-12-26-policy-gradients.md

```diff
@@ -0,0 +1,136 @@
+
+
+In practice, this performs poorly, which I think is for two main reasons:
+
+#### 1. Multiplying every gradient term in a trajectory $$\inline R(\tau)$$ doesn’t discriminate between good harmless actions and bad actions which actually led to bad reward.
+
+R(T) is the sum of rewards which came _after_ the action was taken. So we are trying to form a cause-effect relationship between actions and rewards.
+This kind of return is called **_return-to-go_**. Let's denote the returns recieved in trajectory $$\inline \tau$$ after and including time $$\inline t$$ with $$\inline RTG(\tau, t)$$
+
+
+#### 2. Returns can have high variance. 
+Say, we're in a state where there are 3 possible actions from state $$\inline s$$
+
+action $$\inline a_{1}$$ : return -10
+
+action $$\inline a_{2}$$ : return -60
+
+action $$\inline a_{3}$$ : return -80
+
+and our agent takes action $$\inline a_{1}$$
+
+The agent gets a negative return of -10, so our basic algorithm will end up _discouraging_ this action. But this is not the correct thing to do! $$\inline a_{1}$$ is the best action we can take if we end up in this state, so the algorithm _should encourage_ it. Gotta make lemonade when life gives us lemons, right?
+
+We need to somehow find if an action was _better than average_ among all actions in that state. **_Baselines_** help us achieve this. Instead of using $$\inline R(\tau)$$ as weights, we use $$\inline R(\tau) - b$$. Here,  $$\inline b$$ can be any function, the only constraint is it shouldn’t depend on $$\inline \theta$$. (If it did, the gradient wouldn’t remain the same)
+
+So our gradient term becomes: 
+
+$$\nabla_{\theta} U(\pi)\ \approx {1 \over m}\sum_{i=0}^{m-1}\sum_{t=0}^{H-1} \Big(RTG(\tau, t) - b(s)\Big) \nabla_{\theta}\ log\ \pi (a_{t}|s_{t})$$
+
+Some common choices for b:
+1. Mean return for all $$\inline (s, a) $$ in the current batch. This isn't the best option for b, but this already performs better and is more stable than the basic version. For a sample implementation of this, check out [this notebook on colab](https://colab.research.google.com/drive/1KwHIlm7xesrm3wg0ks-TlvHy09br4PAk)
+2. Time based: $$\inline b(s_{t}) = average\ returns\ after\ s_{t}$$ (in the current batch)
+3. Learned baselines: $$\inline b(s_{t}) = V_{\phi}(s_{t}) $$ (these kinds have a separate critic, which is, in a sense, evaluating the agent’s actions. Discussed next)
+
+
+## Actor-Critic methods
+
+
+These are policy gradient methods where the baseline is also learned.
+Two networks: Critic $$\inline V_{\phi}(s)$$ and Actor $$\inline \pi_{\theta}(a|s)$$
+Both may be separate networks, or may share initial layers and branch out at a later layer.
+
+Critic: did this state turn out more rewarding than I expected?
+
+Actor: did this action send us into a better situation than we thought it would?
+
+If we use this value function as a baseline, our gradient term becomes: 
+
```

Comment:
Here "this value function" isn't very clear. Elaborate more on Actor and Critic.